examples/timeseries/timeseries_traffic_forecasting.py

```diff
@@ -0,0 +1,655 @@
+timeseries
+and predict the future values of each timeseries using the past values of the same
+timeseries.
+This method, however, ignores the dependency of the traffic speed of one road segment on
+the neighboring segments. To be able to take into account the complex interactions between
+the traffic speed on a collection of neighboring roads, we can define the traffic network
+as a graph and consider the traffic speed as a signal on this graph. In this example,
+we implement a neural network architecture which can process timeseries data over a graph.
+We first show how to process the data and create a
+[tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) for
+forecasting over graphs. Then, we implement a model which uses graph convolution and
+LSTM layers to perform forecasting over a graph.
+
+The data processing and the model architecture are inspired by this paper:
+
+Yu, Bing, Haoteng Yin, and Zhanxing Zhu. "Spatio-temporal graph convolutional networks:
+a deep learning framework for traffic forecasting." Proceedings of the 27th International
+Joint
+Conference on Artificial Intelligence. 2018.
+([github](https://github.com/VeritasYin/STGCN_IJCAI-18))
+"""
+
+"""
+## Setup
+"""
+
+import pandas as pd
+import numpy as np
+import os
+from typing import Tuple, Optional
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers, models, optimizers, losses, callbacks, initializers
+
+"""
+## Data preparation
+"""
+
+"""
+### Data description
+"""
+
+"""
+We use a real world traffic speed dataset named `PeMSD7`. We use the version
+collected and prepared by [Yu et al., 2018](https://arxiv.org/abs/1709.04875)
+and available
+[here](https://github.com/VeritasYin/STGCN_IJCAI-18/tree/master/data_loader).
+The data consists of two files: `W_228.csv` contains the distances between 228
+stations across the District 7 of California, and `V_228.csv` contains traffic
+speed collected for those stations in the weekdays of May and June of 2012.
+The full description of the dataset can be found in [Yu et al.,
+2018](https://arxiv.org/abs/1709.04875).
+"""
+
+"""
+### Loading data
+"""
+
+url = "https://github.com/VeritasYin/STGCN_IJCAI-18/raw/master/data_loader/PeMS-M.zip"
+data_dir = keras.utils.get_file(origin=url, extract=True, archive_format="zip")
+data_dir = data_dir.rstrip(".zip")
+
+route_distances = pd.read_csv(
+    os.path.join(data_dir, "W_228.csv"), header=None
+).to_numpy()
+speeds_array = pd.read_csv(os.path.join(data_dir, "V_228.csv"), header=None).to_numpy()
+
+print(f"route_distances shape={route_distances.shape}")
+print(f"speeds_array shape={speeds_array.shape}")
+
+"""
+### sub-sampling roads
+"""
+
+"""
+To reduce the problem size and make the training faster, we will only
+work with a sample of 26 roads out of the 228 roads in the dataset.
+We have chosen the roads by starting from road 0, choosing the 5 closest
```

Comment:
What was the motivation for this selection process? It's not obvious, please explain